src/totp/totp.cpp

```diff
@@ -28,114 +29,107 @@
+                if (vars[1] == STEAM_SHORTNAME) {
+                    // Explicit steam encoder
+                    settings->encoder = steamEncoder();
+                } else {
+                    // Extract step and digits
+                    settings->step = vars[0].toUInt();
+                    settings->digits = vars[1].toUInt();
+                }
             }
+        }
+    }
 
-            q_step = query.queryItemValue("step").toUInt();
-            if (q_step > 0 && q_step <= 60) {
-                step = q_step;
-            }
+    // Bound digits and step
+    settings->digits = qMax(1u, settings->digits);
+    settings->step = qBound(1u, settings->step, 60u);
 
-        } else {
-            seed = key;
-        }
+    // Detect custom settings, used by setup GUI
+    if (settings->encoder.shortName != STEAM_SHORTNAME
+        && (settings->digits != DEFAULT_DIGITS || settings->step != DEFAULT_STEP)) {
+        settings->custom = true;
     }
 
-    if (digits == 0) {
-        digits = defaultDigits;
+    return settings;
+}
+
+QSharedPointer<Totp::Settings> Totp::createSettings(const QString& key, const uint digits, const uint step,
+                                                    const QString& encoderShortName)
+{
+    bool isCustom = digits != DEFAULT_DIGITS || step != DEFAULT_STEP;
+    return QSharedPointer<Totp::Settings>(new Totp::Settings {
+        getEncoderByShortName(encoderShortName), key, false, isCustom, digits, step
+    });
+}
+
+QString Totp::writeSettings(const QSharedPointer<Totp::Settings> settings)
+{
+    // OTP Url output
+    if (settings->otpUrl) {
+        auto urlstring = QString("key=%1&step=%2&size=%3").arg(settings->key).arg(settings->step).arg(settings->digits);
+        if (!settings->encoder.name.isEmpty()) {
+            urlstring.append("&encoder=").append(settings->encoder.name);
+        }
+        return urlstring;
     }
 
-    if (step == 0) {
-        step = defaultStep;
+    // Semicolon output [step];[encoder]
+    if (!settings->encoder.shortName.isEmpty()) {
+        return QString("%1;%2").arg(settings->step).arg(settings->encoder.shortName);
     }
 
-    return seed;
+    // Semicolon output [step];[digits]
+    return QString("%1;%2").arg(settings->step).arg(settings->digits);
 }
 
-QString Totp::generateTotp(const QByteArray key,
-                           quint64 time,
-                           const quint8 numDigits = defaultDigits,
-                           const quint8 step = defaultStep)
+QString Totp::generateTotp(const QSharedPointer<Totp::Settings> settings, const quint64 time)
 {
-    quint64 current = qToBigEndian(time / step);
+    const Encoder& encoder = settings->encoder;
+    uint step = settings->custom ? settings->step : encoder.step;
+    uint digits = settings->custom ? settings->digits : encoder.digits;
+
+    quint64 current;
+    if (time == 0) {
+        current = qToBigEndian(QDateTime::currentDateTime().toTime_t() / step);
```

Comment:
Found the bug!

`toTime_t()` returns a 32bit uint, but it worked with the old code because it was stored in a quint64 variable before being passed to `qToBigEndian`.

`toTime_t()` is also [obsolete](https://doc.qt.io/qt-5/qdatetime-obsolete.html#toTime_t) and we should replace it with `toSecsSinceEpoch()`, which already returns a `qint64`.

Reply:
I'll use the new function, seems the old one is way overdue for deletion.

Reply:
`toSecsSinceEpoch()` was introduced in 5.8 so we would need to increase the minimum Qt version requirement at https://github.com/keepassxreboot/keepassxc/blob/develop/INSTALL.md too

Reply:
ugh...